---
title: "ttBulk Anaylsis"
author: "Isobel Beasley"
date: "Last Updated 10/12/2019"
output: 
html_document: 
  toc: yes
  toc_depth: 3
  number_sections: true
  theme: readable
  highlight: null
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi = 100, 
                      echo=TRUE, 
                      warning=FALSE, message=FALSE, eval = TRUE,
                      fig.show=TRUE, fig.width= 9,fig.height= 8,fig.align='center', out.width = '100%', fig.path= 'Figures/', fig.pos = 'H')
```

# 0- Preparation 

## 0.1 Load Required Packages

```{r, message=FALSE, warning-FALSE}
library(ttBulk)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(forcats)
library(lemon)
```

```{r load prep data, include=FALSE}
load("/stornext/Home/data/allstaff/b/beasley.i/TABI/dev/ttBulk_PCA/PCA_PC_TCGA.rda")
load("/stornext/Home/data/allstaff/b/beasley.i/TABI/dev/ttBulk_PCA/lm_test_nf.rda")
load("/stornext/Home/data/allstaff/b/beasley.i/TABI/dev/ttBulk_PCA/lm_test_PCA.rda")
```

## 0.2 Data Size Reduction 

```{r, eval=FALSE}
#Removed Columns (because they were either duplicates, all NAs or all same value)

ed_PC_TCGA<- PC_TCGA %>% select(-bmi, -weight, -patient, -Sample, -prior_malignancy, -cigarettes_per_day, 
-alcohol_history, -alcohol_intensity,-years_smoked, -height, -iso, -is_recurred, -days_to_recurrence, -Cancer.type, -tissue, -sample.uniqe.1, -sample.uniqe.2, -sample.uniqe.4, -ens, -`merged transcripts`, -sample, -primary_site, 
-site_of_resection_or_biopsy, -X,-ABSOLUTE, -state.x, -state.y, -cases_0_project_disease_type, 
-cases_0_samples_0_sample_type, -classification_of_tumor, -last_known_disease_status, -updated_datetime.x, 
-created_datetime.x, -tissue_or_organ_of_origin, -updated_datetime.y, -created_datetime.y, -`filter out low counts`, 
-created_datetime.y, -days_to_last_known_disease_status, -progression_or_recurrence, -gender, -updated_datetime, -project, -center, -race, -state, -state.1, -created_datetime, -releasable, -dbgap_accession_number, -released, -ethnicity, -disease_type, -name, -is_ffpe, -subtype_patient)

```

## 0.3 ttBulk preparation steps (Adjusting/normalising gene counts already prepared in data)

# 1 - Using ttBulk 

## 1.1 ttBulk PCA Analysis / Dimension Reduction

### 1.1.1 Principal Component Overview and Selection

```{r PCA,eval=FALSE}
#Reduce Dimensions using PCA Analysis (first six PC components)
PCA_PC_TCGA<-ed_PC_TCGA %>% 
  filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F) %>% 
  reduce_dimensions(.element=Sample.ID, .feature=transcript, .abundance=`read count normalised`, method="PCA", .dims=6)
```

```{r, eval=FALSE, include=FALSE}
#Reduce Dimension using PCA Analysis (first six components) 
PCA_PC_TCGA_1<-ed_PC_TCGA %>% 
  filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F) %>% 
  reduce_dimensions(.element=Sample.ID, .feature=transcript, .abundance=`read count normalised`, method="PCA", .dims=1)
```

```{r, echo=FALSE}
PV<-c(0.54998397,0.04981551,0.04178663,0.03221718, 0.02478725, 0.01216323)
barplot(PV, xlab="Principal Components", ylab="Proportion of Variance", main="Proportion of Variance Explained by Principal Components")
```

```{r}
#Proportion of vairance explained by each of the Principal Components
PV
```

*Comment* Above barplot suggests that only the first Principal Component should be utilised in dimension reduction, as very little additional variance explanation is gained by increasing the number of principal components beyond it (shown by the dramatic drop in proportion variance explained by the principal components). Additionaly, the first principal component alone explains a majority of the variance within the population. 

## 1.1.2 What Explains the clustering of data points in PCA analysis? 

```{r}
#Create function for plotting of PCs coloured by categorical groups
plot_PC<- function(x){
	d<-PCA_PC_TCGA<-PCA_PC_TCGA %>% 
		select(contains("PC"), x) %>% 
		distinct() %>% mutate_at(.cols=x, .funs=as.factor)
	h1<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC1", y="PC2", colour=x)) + geom_point(alpha=0.7) + labs(y="PC2 (4.98%)", x="PC1 (55.0%)")
	h2<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC2", y="PC3", colour=x)) + geom_point(alpha=0.7) + labs(y="PC3 (4.18%)",x="PC2 (4.98%)")
	h3<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC3", y="PC4", colour=x)) + geom_point(alpha=0.7) + labs(y="PC4 (3.22%)", x= "PC3 (4.18%)")
	h4<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC4", y="PC5", colour=x)) + geom_point(alpha=0.7) + labs(y="PC5 (2.48%)", x="PC4 (3.22%)")
	h5<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC5", y="PC6", colour=x)) + geom_point(alpha=0.7) + labs(y="PC6 (1.22%", x="PC5 (2.48%)")
	plot<-grid_arrange_shared_legend(h1,h2, h3, h4, h5, ncol=3, nrow=2)
	return(plot)
}

#Create a function for plotting of PCs coloured by continous data
plot_PC_1<- function(x){
	d<-PCA_PC_TCGA<-PCA_PC_TCGA %>% 
		select(contains("PC"), x) %>% 
		distinct()
	h1<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC1", y="PC2", colour=x)) + geom_point(alpha=0.7) + labs(y="PC2 (4.98%)", x="PC1 (55.0%)")
	h2<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC2", y="PC3", colour=x)) + geom_point(alpha=0.7) + labs(y="PC3 (4.18%)",x="PC2 (4.98%)")
	h3<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC3", y="PC4", colour=x)) + geom_point(alpha=0.7) + labs(y="PC4 (3.22%)", x= "PC3 (4.18%)")
	h4<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC4", y="PC5", colour=x)) + geom_point(alpha=0.7) + labs(y="PC5 (2.48%)", x="PC4 (3.22%)")
	h5<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC5", y="PC6", colour=x)) + geom_point(alpha=0.7) + labs(y="PC6 (1.22%", x="PC5 (2.48%)")
	plot<-grid_arrange_shared_legend(h1,h2, h3, h4, h5, ncol=3, nrow=2)
	return(plot)
}
```

### (A) Participant Features

### A.1 Race Analysis 

```{r}
c <- PCA_PC_TCGA %>% select(Sample.ID, contains("PC"), subtype_Age, subtype_Race, age_at_diagnosis) %>% distinct() 

#For race comparasion, make NA's a new factor; "Unknown"
c$subtype_Race<-fct_explicit_na(c$subtype_Race, na_level="Unknown")
#Make Black_or_African_American and Black or African American a single factor level
c$subtype_Race<-fct_collapse(c$subtype_Race, "BLACK_OR_AFRICAN_AMERICAN"=c("BLACK OR AFRICAN AMERICAN", "BLACK_OR_AFRICAN_AMERICAN"))
PV

#Plot PCs against each other colouring by race
p1<-c %>% ggplot(aes(x=PC1, y=PC2, colour=subtype_Race)) + geom_point(alpha=0.7) + labs(y="PC2 (4.98%)", x="PC1 (55.0%)")
p2<- c %>% ggplot(aes(x=PC2, y=PC3, colour=subtype_Race)) + geom_point(alpha=0.7) +labs(y="PC3 (4.18%)",x="PC2 (4.98%)")
p3<- c %>% ggplot(aes(x=PC3, y=PC4, colour=subtype_Race)) + geom_point(alpha=0.7) + labs(y="PC4 (3.22%)", x= "PC3 (4.18%)")
p4<- c %>% ggplot(aes(x=PC4, y=PC5, colour=subtype_Race)) + geom_point(alpha=0.7) + labs(y="PC5 (2.48%)", x="PC4 (3.22%)")
p5<- c %>% ggplot(aes(x=PC5, y=PC6, colour=subtype_Race)) + geom_point(alpha=0.7) + labs(y="PC6 (1.22%", x="PC5 (2.48%)")


grid_arrange_shared_legend(p1,p2, p3, p4, p5, ncol=3, nrow=2) 
```
*Comment* There appears to be no clustering of indivdiuals into ethnic groups, although a lack of recording (high number of unknown data points) may be hidding potetional clustering. 

### A.2 Age analysis

```{r, results="hide"}
#Age Comparasion
age<-c("age_at_diagnosis", "year_of_birth", "subtype_Age")
sapply(age, FUN=plot_PC_1)
```

*Comment* As seen above there appears to little or no clustering of data points into age at diagnosis, subtype_Age or year of birth. This suggests that these factors do not explain are large proportion of the variance inbetween individuals in this sample. 

### A.3 Time to death/Survival Status/Disease Free Status
```{r,  results="hide"}
sapply(c("DFS_STATUS", "vital_status"), plot_PC)
sapply(c("DFS_MONTHS", "days_to_death"), plot_PC_1)
```

*Comment* There appears to be no clustering of data/individuals by survival status, disease free status etc. Of particular note, in vital_status those small number of individuals with a recurrence are not found clustered together. 

### (B) Clinical Features

```{r, results="hide", fig.height=20}
plot_PC(names(PCA_PC_TCGA)[35])
```

```{r, results="hide"}
sapply(c(names(PCA_PC_TCGA)[36:43]), FUN=plot_PC)
```

*Comment* The above graphs indicicate that there is clustering of individuals by Gleason score, which is stronger for reviewed Gleason scores than for clinical Gleason scores. To a lesser extent there appears to be clustering of data point also by Tumour_cellularity_pathology. 

### (C) Cellular/Tumour Morphology

```{r, results="hide"}
sapply(c("purity.score", "tumor_stage", "morphology", "Type", "primary_diagnosis"), FUN=plot_PC_1)
```

*Comment* Of the above graphs, there appears to only be clustering of individuals by purity score (and not any of the other above measurements)

### Molecular Markers

### (D) Categorical Molecular Markers

```{r,results="hide"}
mol_mark<-names(PCA_PC_TCGA)[53:112]
ex<-c("subtype_exon_imbalance_score", "subtype_AR_score", "subtype_AR_protein", "subtype_AR_V7_ratio", "subtype_Fraction_genome_altered", "subtype_AR_mRNA")
mol_mark<-setdiff(mol_mark, ex)
sapply(mol_mark, FUN=plot_PC)
```

Above graphs indiciate theres is clustering of samples by subtype_iCluster, mRNA cluster, methylation cluster, ERG_Status, SPOP mutation Status, and TP53_CNA. This is also true to a lesser extent of miRNA cluster, RPRA Cluster, PTEN_CNA  and SPOP_CNA (See repeated plots below;) 

```{r, echo=FALSE, results="hide"}

res<- function(x){
	d<-PCA_PC_TCGA<-PCA_PC_TCGA %>% 
		select(contains("PC"), x) %>% 
		distinct() %>% mutate_at(.cols=x, .funs=as.factor)
	h1<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC1", y="PC2", colour=x)) + geom_point(alpha=0.7)
	return(h1)
}

test<-c("subtype_iCluster", "subtype_methylation_cluster", "subtype_SPOP_mut", "subtype_PTEN_CNA", "subtype_TP53_CNA")
sapply(test, FUN=res)
```

### (E) Continuous Features

```{r, include=FALSE}
plot_PC_2<- function(x){
	d<-PCA_PC_TCGA<-PCA_PC_TCGA %>% 
		select(contains("PC"), x) %>% 
		distinct() %>% mutate_at(.cols=x, .funs=as.numeric)
	h1<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC1", y="PC2", colour=x)) + geom_point(alpha=0.7) + labs(y="PC2 (4.98%)", x="PC1 (55.0%)")
	h2<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC2", y="PC3", colour=x)) + geom_point(alpha=0.7) + labs(y="PC3 (4.18%)",x="PC2 (4.98%)")
	h3<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC3", y="PC4", colour=x)) + geom_point(alpha=0.7) + labs(y="PC4 (3.22%)", x= "PC3 (4.18%)")
	h4<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC4", y="PC5", colour=x)) + geom_point(alpha=0.7) + labs(y="PC5 (2.48%)", x="PC4 (3.22%)")
	h5<- PCA_PC_TCGA %>% ggplot(aes_string(x="PC5", y="PC6", colour=x)) + geom_point(alpha=0.7) + labs(y="PC6 (1.22%", x="PC5 (2.48%)")
	plot<-grid_arrange_shared_legend(h1,h2, h3, h4, h5, ncol=3, nrow=2)
	return(plot)
}
```

```{r, results="hide", fig.height=15}
sapply(c("subtype_exon_imbalance_score", "subtype_AR_score", "subtype_AR_protein", "subtype_AR_V7_ratio", "subtype_AR_V7_reads", "subtype_Fraction_genome_altered", names(PCA_PC_TCGA)[44:52]), FUN=plot_PC_2)
```

###  (F) Unknown Etc. 

```{r, fig.height=15, results="hide"}
cat<-c("vial", "plate", "hg")
sapply(cat, plot_PC)
cont<-c("LUMP","CPE", "IHC", "ESTIMATE")
sapply(cont, plot_PC_2)
```

*Comment* There appears to be no clustering by any of the categorical features above

## Summary (1.1.2)

## 1.3 Test differential abundance 

### Focusing on those which demonstrated clustering above

```{r, eval=FALSE}
ed_PC_TCGA$`read count`

#Differential Abundance subtype_iCluster
e4<-PCA_PC_TCGA %>% 
filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F, is.na(subtype_iCluster)==F) %>%
test_differential_abundance(.formula= ~subtype_iCluster, .sample=Sample.ID, .transcript=transcript, .abundance = `read count normalised`, action="get")

e1<-ed_PC_TCGA %>% 
filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F, is.na(subtype_iCluster)==F) %>%
test_differential_abundance(.formula= ~subtype_iCluster, .sample=Sample.ID, .transcript=transcript, .abundance = `read count`, action="get")

ed_PC_TCGA$subtype_methylation_cluster

e<-ed_PC_TCGA %>% 
filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F, is.na(subtype_methylation_cluster)==F) %>%
test_differential_abundance(.formula= ~subtype_methylation_cluster, .sample=Sample.ID, .transcript=transcript, .abundance = `read count normalised`, action="get")

e<-ed_PC_TCGA %>% 
filter(is.na(transcript)==F, is.na(Sample.ID)==F, is.na(`read count normalised`)==F, is.na(subtype_methylation_cluster)==F) %>%
test_differential_abundance(.formula= ~subtype_methylation_cluster, .sample=Sample.ID, .transcript=transcript, .abundance = `read count normalised`, action="get")

hist(e$PValue)

sum(is.na(row.names(ed_PC_TCGA)))


Test<-PCA_PC_TCGA %>% 
	select(Sample.ID, `read count normalised`, `CAPRA-S`, subtype_Race, subtype_Age, subtype_iCluster, subtype_mRNA_cluster, subtype_miRNA_cluster, subtype_methylation_cluster, subtype_ERG_status, subtype_PTEN_CNA, subtype_RB1_CNA, subtype_TP53_CNA, age_at_diagnosis, year_of_birth, year_of_death) %>% 
	group_by(Sample.ID) %>% 
	mutate(mean(`read count normalised`)) %>% 
	ungroup () %>% 
	select(-`read count normalised`, -Sample.ID) %>% 
	distinct()

lm_test_data<-ed_PC_TCGA %>% 
	select(Sample.ID, `read count normalised`, `CAPRA-S`, subtype_Race, subtype_Age, subtype_iCluster, subtype_mRNA_cluster, subtype_miRNA_cluster, subtype_methylation_cluster, subtype_ERG_status, subtype_PTEN_CNA, subtype_RB1_CNA, subtype_TP53_CNA, age_at_diagnosis, year_of_birth, year_of_death) 



```

## 1.4 Does introducing Principal Components as covariates remove biological information? 

```{r, eval=FALSE}
#Unfiltered data set, simple linear regression
#with PC1 added as a covariate
lm_test_PCA<- PCA_PC_TCGA_1 %>%
  group_by(transcript) %>% 
  summarise(
    beta = summary(lm(log(`read count normalised`+1)~ `CAPRA-S`+PC1))$coefficients[2,1],
    pval= summary(lm(log(`read count normalised`+1)~`CAPRA-S`+PC1))$coefficients[2,4])
```

```{r}
#Number of Genes detected as being significantly associated with CAPRA-S risk
#When PC1 is added as a covariate
sum(lm_test_PCA<0.05, na.rm=T)
```

```{r, eval=FALSE}
#Unfiltered data set simple linear regression (no covariate)
lm_test_nf <-  PCA_PC_TCGA_1 %>%
  group_by(transcript) %>% 
  summarise(
    beta = summary(lm(log(`read count normalised`+1)~ `CAPRA-S`))$coefficients[2,1],
    pval= summary(lm(log(`read count normalised`+1)~`CAPRA-S`))$coefficients[2,4])
```

```{r}
#Number of Genes detected as being significantly associated with CAPRA-S risk (no covariate)
sum(lm_test_nf<0.05, na.rm=T)
```

```{r}
lm_nf<- lm_test_nf %>% ggplot(aes(x=pval)) +geom_histogram(colour="dodgerblue", fill="white", na.rm=T) + ylim(0, 2500) +labs(title="No Covariate")
lm_PCA<- lm_test_PCA %>% ggplot(aes(x=pval)) +geom_histogram(colour="dodgerblue", fill="white", na.rm=T) + ylim(0, 2500) +labs(title="PC Component 1 as Covariate")
grid.arrange(lm_nf, lm_PCA, ncol=2)
```

## 1.5 Does introducing known confounders remove biological information (Covariate Analysis Rmd Document)? 

